File: WheelsCodingChallenge/Controller/AddUserVC.m
//
//  AddUserVC.m
//  WheelsCodingChallenge
//
//  Created by Mike Saradeth on 4/4/19.
//  Copyright © 2019 Mike Saradeth. All rights reserved.
//

#import "AddUserVC.h"

@interface AddUserVC ()
@end

@implementation AddUserVC
@synthesize tableView;
@synthesize items;


- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupVC];
    items = [TextFieldItem loadData];
}


#pragma mark - setup VC
- (void)setupVC {
    tableView.dataSource = self;
    [tableView registerNib:[UINib nibWithNibName:@"AddUserCell" bundle:nil] forCellReuseIdentifier:@"Cell"];
}


#pragma mark - Add New User
- (IBAction)addUserAction:(id)sender {
    User *user = [self getUser];
    [self.delegate addNewUser:user];
    [self.navigationController popViewControllerAnimated:true];
}
- (NSNumber*) getItem:(int)index {
    TextFieldItem *item = items[index];
    NSNumber *number = @([item.text intValue]);
    return  number;
}



#pragma mark - UITableViewDataSource, UITextFieldDelegate
- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    AddUserCell *cell =  [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    cell.textField.delegate = self;
    [cell configure: items[indexPath.row] row:indexPath.row];
    return cell;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return items.count;
}

- (void) textFieldDidEndEditing:(UITextField *)textField {
    ((TextFieldItem*)items[textField.tag]).text = textField.text;
}


#pragma mark - Helper methods
- (User*)getUser {
    Badge *badge = [[Badge alloc] initWithData:[self getItem:2]
                                        silver:[self getItem:3]
                                        bronze:[self getItem:4]];
    User *user = [[User alloc] initWithData:((TextFieldItem*)items[0]).text
                                  urlString:@"personPlaceholder"
                                 reputation:[self getItem:1]
                                      badge:badge];
    return user;
}


@end

File: WheelsCodingChallenge/View/AddUserVC.swift
//
//  AddUserVC.swift
//  WheelsCodingChallenge
//
//  Created by Mike Saradeth on 4/2/19.
//  Copyright © 2019 Mike Saradeth. All rights reserved.
//

import UIKit

class AddUserVC: UIViewController {
    @IBOutlet weak var tableView: UITableView!
    var viewModel: AddUserViewModel!
    
    static func createWith(title: String, viewModel: AddUserViewModel) -> AddUserVC {
        let storyboard = UIStoryboard.init(name: "StackUserList", bundle: nil)
        let vc = storyboard.instantiateViewController(withIdentifier: "AddUserVC") as! AddUserVC
        vc.title = title
        vc.viewModel = viewModel
        return vc
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        tableView.delegate = self
        tableView.dataSource = self
        tableView.register(UINib(nibName: "AddUserTableViewCell", bundle: nil), forCellReuseIdentifier: AddUserTableViewCell.cellIdentifier)
    }
    
  
    
    @IBAction func addUserAction(_ sender: Any) {
        navigationController?.popViewController(animated: true)
    }
    
}

extension AddUserVC: UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let cell = tableView.dequeueReusableCell(withIdentifier: AddUserTableViewCell.cellIdentifier, for: indexPath) as! AddUserTableViewCell
        cell.configure(item: viewModel.user)
        
        return cell
    }
    
}

extension AddUserVC: UITableViewDelegate {
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return UITableView.automaticDimension
    }
}

